### PR TITLE
API for KEP-4831: Control VPA eviction behavior based on scaling direction and resource

### PR DESCRIPTION
**Pull request by voelzmo**
#### Which component this PR applies to?

vertical-pod-autoscaler

#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it
Implements KEP-4831: Control VPA eviction behavior based on scaling direction and resource

#### Which issue(s) this PR fixes:

API for #4831
Implementation is done in #5599

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by voelzmo** [comment]
While looking for the best place to put this into the Updater, I found https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/priority/pod_eviction_admission.go which was pulled in years ago with https://github.com/kubernetes/autoscaler/pull/1002 but apparently never used afterwards. Beata (after merging) even commented that the default implementation isn't even used: https://github.com/kubernetes/autoscaler/pull/1002#pullrequestreview-130869322

I guess this is probably a good place to implement the filtering we want to do based on the comparison of `target` and `requests` and if no restrictions are given, just use the already implemented `noopPodEvictionAdmission`?

**Comment by jbartosik** [comment]
> While looking for the best place to put this into the Updater, I found https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/priority/pod_eviction_admission.go which was pulled in years ago with #1002 but apparently never used afterwards. Beata (after merging) even commented that the default implementation isn't even used: [#1002 (review)](https://github.com/kubernetes/autoscaler/pull/1002#pullrequestreview-130869322)
> 
> I guess this is probably a good place to implement the filtering we want to do based on the comparison of `target` and `requests` and if no restrictions are given, just use the already implemented `noopPodEvictionAdmission`?

Yes, adding a new type that implements `PodEvictionAdmission` and passing it to `Updater` seems to be the best way to implement this KEP.

I'd implement it so when there are no restrictions it just allows the pod

> 
> There's also https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go but that seems to be concerned with limiting eviction based on minimum available replicas and ensuring this also works for Jobs and other Pod owners in the same way.
> 
> Joachim WDYT, should I just make use of the PodEvictionAdmission linked above? From a flow perspective, this seems to be a good fit, [as it is used in the priority calculator](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator.go#L156), where Pod requests and target recommendation are compared to figure out if a Pod should be evicted and in which order.

Yes,

**Comment by voelzmo** [comment]
@jbartosik Can you please take a look? Thanks!

**Comment by voelzmo** [comment]
@jbartosik Can you try to give this a review, please? Thanks!

**Review by jbartosik — COMMENTED** [review]
Since in-place pod updates are now finally a thing (alpha but I hope for beta soon) I'd like to make sure this works well with support for in place updates.

I can see two approaches:

- have one "actuationRequirements" that affect both in-place and eviction actuation
- have separate controls for "evictionRequirements" and "inPlaceUpdateRequirements"

I lean towards single "actuationRequirements" because it's simpler and I don't see how I'd use separate controls for eviction and in place update. What do you think?

File: vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ kind: CustomResourceDefinition
 metadata:
   annotations:
     api-approved.kubernetes.io: https://github.com/kubernetes/kubernetes/pull/63797
-    controller-gen.kubebuilder.io/version: v0.4.0
```
[the remark this answers is not in the document]
Comment (jbartosik) [reply vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml]: > 

I sent https://github.com/kubernetes/autoscaler/pull/5754

**Comment by jbartosik** [comment]
Let's decide if we want to change the API for in-place updates before merging this (https://github.com/kubernetes/autoscaler/pull/5754)

**Comment by voelzmo** [comment]
Force-pushed an update:
* rebased on current master
* removed the `OrEqual` ChangeRequirements as discussed in https://github.com/kubernetes/autoscaler/pull/5754#issuecomment-1647577540

File: vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml [hunk 2]
```diff
@@ -379,6 +379,38 @@ spec:
...
+                            current requests, what kind of change is necessary for
+                            the Pod to be evicted
+                          enum:
+                          - TargetHigherThanRequests
+                          - TargetLowerThanRequests
+                          type: string
+                        resource:
+                          description: Resources is a list of one or more resources
+                            that the condition applies to. If more than one resource
+                            is given, they are combined with OR, not with AND.
```
Comment (jbartosik) [inline vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml:403]:
```suggestion
                            is given, requirement is fulfilled if at least one resource meets `changeRequirement`
```

**Review by jbartosik — COMMENTED** [review]
Looks good to me

@mwielgus @gjtempleton could you take a look please

File: vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml (hunk 2 above)
Comment (voelzmo) [reply vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml]: done, thanks!

**Review by mwielgus — APPROVED** [review]
/lgtm
/approve

**Pull request merged**